src/common/io.py

```diff
@@ -1457,13 +1459,12 @@ def _get_trajectory(self, data_index, start_index = 0, stop_index = None):
     def _get_diagnostics_trajectory(self, data_index, start_index = 0, stop_index = None):
         """ Returns trajectory for the diagnostics variable that corresponds to index 'data_index'. """
         self._verify_file_data()
-
-        index_in_cache = data_index in self._data_3
-        partial_cache_ok = self._can_use_partial_cache(start_index, stop_index)
-        if (index_in_cache and not self._allow_file_updates) or (index_in_cache and partial_cache_ok):
-            return self._data_3[data_index]
-        self._data_3[data_index] = self._read_trajectory_data(data_index, True, start_index, stop_index)[start_index:stop_index]
-        return self._data_3[data_index]
+        # caching
+        current_data = self._data_3.get(data_index, None)
+        if (current_data is not None) and self._can_use_partial_cache(current_data, stop_index):
```

Comment:
Since this `if ...` is written now in several places I suggest we break it into a separate function, I mean only the expression being evaluated, that is:
```
if f(x):
   ...
```

Reply:
Fixed; please have a look. 
Kept some extra parenthesis for readability though.